src/cpu/pred/simple_btb.hh

```diff
@@ -52,41 +54,137 @@ namespace gem5
 namespace branch_prediction
 {
 
-class SimpleBTB : public BranchTargetBuffer
+class BTBEntry : public CacheEntry
 {
   public:
-    SimpleBTB(const SimpleBTBParams &params);
+    /** The entry's tag. */
+    Addr tag;
 
-    void memInvalidate() override;
-    bool valid(ThreadID tid, Addr instPC) override;
-    const PCStateBase *lookup(ThreadID tid, Addr instPC,
-                           BranchType type = BranchType::NoBranch) override;
-    void update(ThreadID tid, Addr instPC, const PCStateBase &target_pc,
-                           BranchType type = BranchType::NoBranch,
-                           StaticInstPtr inst = nullptr) override;
-    const StaticInstPtr getInst(ThreadID tid, Addr instPC) override;
+    /** The entry's target. */
+    std::unique_ptr<PCStateBase> target;
 
+    /** The entry's thread id. */
+    ThreadID tid;
 
+    /** Whether or not the entry is valid. */
+    bool valid;
+
+    /** Pointer to the static branch instruction at this address */
+    StaticInstPtr inst;
+
+    /** Update the entry in the BTB */
+    void update(ThreadID _tid,
+                const PCStateBase &_target,
+                StaticInstPtr _inst)
+    {
+        tid = _tid;
+        set(target, _target);
+        inst = _inst;
+        valid = true;
+    }
+
+    /** Match the tag of the BTB entry */
+    bool matchTag(const Addr _tag, ThreadID _tid)
+    {
+        return (isValid() && (tag == _tag) && (tid == _tid));
+    }
+
+    /** Set the tag of the BTB entry */
+    void setTag(Addr _tag) override { tag = _tag; }
+
+    /** Copy constructor */
+    BTBEntry(const BTBEntry &other)
+    {
+        tag   = other.tag;
+        tid   = other.tid;
+        valid = other.valid;
+        inst  = other.inst;
+    }
+
+    /** Default constructor */
+    BTBEntry() : CacheEntry(), tag(0), tid(0), valid(false), inst(nullptr)  {}
+
+    using CacheEntry::matchTag;
+};
+
+class BTBCache : public AssociativeCache<BTBEntry>
+{
   private:
-    struct BTBEntry
+    typedef replacement_policy::Base BaseReplacementPolicy;
+
+    BTBIndexingPolicy *indexingPolicy;
+
+  public:
+    BTBCache(const char *_my_name, const size_t num_entries,
+             const size_t associativity,
+             BaseReplacementPolicy *_replPolicy,
+             BaseIndexingPolicy *_indexingPolicy)
+        : AssociativeCache(_my_name, num_entries, associativity,
+                       _replPolicy, _indexingPolicy, BTBEntry())
+    {
+        indexingPolicy = static_cast<BTBIndexingPolicy*>(_indexingPolicy);
+    }
+
+    BTBEntry *findEntry(const Addr instPC, ThreadID tid)
```

Comment:
We are duplicating a lot of the code in the associative cache just because (it seems) the thread id is part of the address. All of these overrides would be avoided if we used a Tag type instead in the associative cache instead.